src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -551,5 +550,49 @@ private static bool HasTools(string file)
         {
             return file.StartsWith("tools/", StringComparison.OrdinalIgnoreCase);
         }
+
+        /// <summary>
+        /// An <see cref="IEqualityComparer{T}" /> that compares <see cref="LockFileTargetLibrary" /> objects by the value of the <see cref="LockFileTargetLibrary.Name" /> and <see cref="LockFileTargetLibrary.Version" /> properties.
+        /// </summary>
+        private class LockFileTargetLibraryNameAndVersionEqualityComparer : IEqualityComparer<LockFileTargetLibrary>
+        {
+            /// <summary>
+            /// Gets a static singleton for the <see cref="LockFileTargetLibraryNameAndVersionEqualityComparer" /> class.
+            /// </summary>
+            public static LockFileTargetLibraryNameAndVersionEqualityComparer Instance = new();
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="LockFileTargetLibraryNameAndVersionEqualityComparer" /> class.
+            /// </summary>
+            private LockFileTargetLibraryNameAndVersionEqualityComparer()
+            {
+            }
+
+            /// <summary>
+            /// Determines whether the specified <see cref="LockFileTargetLibrary" /> objects are equal by comparing their <see cref="LockFileTargetLibrary.Name" /> and <see cref="LockFileTargetLibrary.Version" /> properties.
+            /// </summary>
+            /// <param name="x">The first <see cref="LockFileTargetLibrary" /> to compare.</param>
+            /// <param name="y">The second <see cref="LockFileTargetLibrary" /> to compare.</param>
+            /// <returns><c>true</c> if the specified <see cref="LockFileTargetLibrary" /> objects' <see cref="FileSystemInfo.Name" /> and <see cref="LockFileTargetLibrary.Version" /> properties are equal, otherwise <c>false</c>.</returns>
+            public bool Equals(LockFileTargetLibrary x, LockFileTargetLibrary y)
+            {
+                return string.Equals(x.Name, y.Name, StringComparison.Ordinal) && x.Version.Equals(y.Version);
```

Comment:
package ids are not case sensitive. Does `LockFileTargetLibrary` only contain lower case names?

Reply:
Looking at the following unit tests which gives me the impression that `LockFileTargetLibrary.Name` property is case sensitive. 

https://github.com/NuGet/NuGet.Client/blob/ba01cc633f88fb2dd1bfcbcdd859fb89525a56eb/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/LockFileTargetLibraryTests.cs#L15-L40

https://github.com/NuGet/NuGet.Client/blob/ba01cc633f88fb2dd1bfcbcdd859fb89525a56eb/test/NuGet.Core.Tests/NuGet.ProjectModel.Test/LockFileTargetLibraryTests.cs#L503-L528

This behavior which matches the current implementation i.e.[ `var libraryKey = library.Name + " " + library.Version;`](https://github.com/NuGet/NuGet.Client/blob/ba01cc633f88fb2dd1bfcbcdd859fb89525a56eb/src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs#L341). The hascode of `libraryKey` will be different if library name differs just by case. [`LockFileTargetLibrary.GetHashCode`](https://github.com/NuGet/NuGet.Client/blob/ba01cc633f88fb2dd1bfcbcdd859fb89525a56eb/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileTargetLibrary.cs#L93) also invokes `combiner.AddObject(Name);` instead of `combiner.AddStringIgnoreCase(Name);`

As per [docs](https://docs.microsoft.com/en-us/dotnet/csharp/how-to/compare-strings#default-ordinal-comparisons), `String.Equals` performs a case-sensitive ordinal comparison. I added `StringComparison.Ordinal` argument to make it more explicit. Hence, I think there is no change to functionality.